.github/workflows/commits.yml

```diff
@@ -0,0 +1,32 @@
+# yaml-language-server: $schema=https://json.schemastore.org/github-workflow.json
+name: Conventional commit check
+on: [push]
+jobs:
+  commit_messages:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Conventional commit check
+        id: conventional_commit_check
+        uses: cocogitto/cocogitto-action@v3
+        with:
+          check-latest-tag-only: true
+
+      - name: Check scope
+        id: check_scopes
+        run: |
+          scopes_file=".github/.scopes.txt"
+          scopes=$(awk 'NR == 1 { printf $1 } NR > 1 { printf "|" $1 }' $scopes_file)
+          regex="(\(($scopes)\):|[^()]:)?"
```

Comment:
Ok, so I will leave a comment quickly explaining this regex. It works like this:

If there is no parenthesis, it matches the colon and a previous character.
If there are parenthesis, what is inside MUST be one of the scopes.

The check fails if it doesn't match anything.

Now, this is susceptible to being dodged, e.g. if some writes `feat(askdhaskd: Changed the whole code`, however I thought to myself that trying to generate the perfect regex for all edge cases would be more time consuming than simply looking at the commits and correcting as needed.

If everyone thinks that a perfect regex is more preferable, I will continue iterating on it.

Reply:
But filtering out this commits would need another regex, right? I would need to check each to commit to see if there's a scope (first solution that comes to mind is checking if there are parenthesis), and run to see if the scope inside is valid.

I don't know if this would simplify the logic by adding another layer of verification. Unless there's something I'm missing, which is possible, I'm not an expert on command line scripts, regex and git manipulation hehehe

Reply:
Yes, but it would be a simpler regex. Something like `with_scopes=$(echo -n "$commits" | grep -e '(.*)!\?:')`.

You could write this script in Lua; it will probably be much easier to understand and maintain. Here is a draft from chat.

```
function extractScopesFromFile(filename)
  local scopes = {}
  local file = io.open(filename, "r")
  if not file then
    error("Failed to open file: " .. filename)
  end
  for line in file:lines() do
    local scope = line:match("%S+")
    if scope then
      scopes[scope] = true
    end
  end
  file:close()
  return scopes
end

function extractScopeFromCommit(commitTitle)
  local scope = commitTitle:match("%a+%((.-)%)")
  return scope
end

local filename = ".github/.scopes.txt"
local scopes = extractScopesFromFile(filename)

for line in io.lines() do
  local commitScope = extractScopeFromCommit(line)
  if commitScope and not scopes[commitScope] then
    print("Invalid scope:", commitScope)
    print("Commit:", line)
    os.exit(1)
  end
end
```

Reply:
I asked chat to convert this script to bash but it is much harder to understand.

Reply:
Another version calling `git log` from inside Lua:
```lua
-- Get commit titles using git log command
local handle = io.popen('git log --pretty=format:"%s" $(git merge-base HEAD origin/main)..HEAD')
local commits = handle:read("*a")
handle:close()

-- Check if the scope matches for each commit
for line in commits:gmatch("[^\r\n]+") do
  local commitScope = extractScopeFromCommit(line)
  if commitScope and not scopes[commitScope] then
    print("Invalid scope:", commitScope)
    print("Commit:", line)
    os.exit(1)
  end
end
```